cardano-api/src/Cardano/Api/IO.hs

```diff
@@ -0,0 +1,41 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE KindSignatures #-}
+
+module Cardano.Api.IO
+  ( File(..)
+  , In
+  , Out
+  , InOut
+  , HasFileMode(..)
+  ) where
+
+import           Data.Kind (Type)
+import           Data.String (IsString)
+
+-- | Phantom type to indicate a file for reading.
+data In
+
+-- | Phantom type to indicate a file for writing.
+data Out
+
+-- | Phantom type to indicate a file for reading or writing.
+data InOut
```

Comment:
Introduce phantom types to track what direction the file will be processed.  An `In` file is only read.  An `Out` file is only written.  An `InOut` file is both read and written.

Reply:
Converted to use `DataKinds`.  This means we now use `'In/'Out/'InOut` instead of 'In/Out/InOut`.